src/main/kotlin/org/rust/ide/refactoring/extractTrait/RsExtractTraitHandler.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.refactoring.extractTrait
+
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.refactoring.RefactoringActionHandler
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.ui.RefactoringDialog
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import com.intellij.ui.components.JBTextField
+import com.intellij.ui.layout.panel
+import com.intellij.util.ui.JBUI
+import org.jetbrains.annotations.TestOnly
+import org.rust.ide.refactoring.RsMemberInfo
+import org.rust.ide.refactoring.RsMemberSelectionPanel
+import org.rust.ide.refactoring.isValidRustVariableIdentifier
+import org.rust.lang.core.psi.RsImplItem
+import org.rust.lang.core.psi.ext.RsItemElement
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.childrenOfType
+import org.rust.openapiext.addTextChangeListener
+import org.rust.openapiext.isUnitTestMode
+import javax.swing.JComponent
+import javax.swing.JTextField
+
+class RsExtractTraitHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        val offset = editor.caretModel.offset
+        val element = file.findElementAt(offset) ?: return
+        val impl = element.ancestorOrSelf<RsImplItem>() ?: return
+        if (impl.traitRef != null) return
+        if (!CommonRefactoringUtil.checkReadOnlyStatus(project, impl)) return
+
+        val members = (impl.members ?: return).childrenOfType<RsItemElement>()
+        if (members.isEmpty()) return
+        val memberInfos = members.map { RsMemberInfo(it, false) }
+
+        val dialog = RsExtractTraitDialog(project, impl, memberInfos)
+        if (isUnitTestMode) {
+            dialog.doAction()
+        } else {
+            dialog.show()
+        }
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        /* not called from the editor */
+    }
+}
+
+class RsExtractTraitDialog(
+    project: Project,
+    private val impl: RsImplItem,
+    private val memberInfos: List<RsMemberInfo>,
+) : RefactoringDialog(project, false) {
+
+    private val traitNameField: JTextField = JBTextField().apply {
+        addTextChangeListener { validateButtons() }
+    }
+
+    init {
+        super.init()
+        title = "Extract Trait"
```

Comment:
I found several `"Extract Trait"` strings in the code. Maybe reuse `action.Rust.RsExtractTrait.text` instead?